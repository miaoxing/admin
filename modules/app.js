import { createRoot } from 'react-dom/client';

import './bootstrap';
import 'antd/dist/reset.css';
import '../scss/index.scss';

import AppContainer from '../components/AppContainer';

const container = document.getElementById('root');
const root = createRoot(container);
root.render(<AppContainer/>);
