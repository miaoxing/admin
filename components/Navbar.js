import {Component} from 'react';
import {history} from '@mxjs/app';
import api from '@mxjs/api';
import $ from 'miaoxing';
import {Layout, Menu, Dropdown} from 'antd';
import {DownOutlined} from '@ant-design/icons';
import {Link} from '@mxjs/router';
import {Actions} from '@mxjs/actions';
import {LockOutlined, LogoutOutlined} from '@ant-design/icons';
import propTypes from 'prop-types';
import Avatar from './Avatar';
import {css, createStyle} from '@mxjs/css';
import {Box} from '@mxjs/box';

export default class extends Component {
  static defaultProps = {
    user: {},
  };

  static propTypes = {
    user: propTypes.object,
  };

  handleLogout = async () => {
    const {ret} = await api.post('logout');
    await $.ret(ret);
    if (ret.isSuc()) {
      window.localStorage.removeItem('token');
      history.push($.url('admin/login', {next: window.location.pathname}));
    }
  };

  render() {
    const {user} = this.props;

    return (
      <Layout.Header
        className={css('toRight', 'zIndex-1', {shadow: '0 1px 2px 0 rgba(0, 0, 0, 0.05)'})}
        /* 开发时，避免 reload 之后被原有样式覆盖 */
        style={createStyle('bgWhite')}
      >
        <Dropdown
          overlay={
            <Menu>
              <Menu.Item>
                <Link to={$.url('admin/password')}>
                  <LockOutlined/>{' '}修改密码
                </Link>
              </Menu.Item>
              <Menu.Item>
                <a onClick={this.handleLogout}>
                  <LogoutOutlined/>{' '}退出登录
                </a>
              </Menu.Item>
            </Menu>
          }
        >
          <Box cursorPointer>
            <Actions>
              <Avatar user={user}/>
              {user.username}
              <DownOutlined/>
            </Actions>
          </Box>
        </Dropdown>
      </Layout.Header>
    );
  }
}
