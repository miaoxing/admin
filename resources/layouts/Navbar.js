import React from "react";
import app, {history} from '@miaoxing/app';
import api from '@miaoxing/api';
import $ from 'miaoxing';
import {Layout, Avatar, Menu, Dropdown} from 'antd';
import {DownOutlined} from '@ant-design/icons';
import {Link} from '@miaoxing/router';
import {Box} from 'rebass';
import {Actions} from '@miaoxing/actions';
import {css} from 'emotion';

export default class extends React.Component {
  static defaultProps = {
    user: {},
  };

  handleLogout = async () => {
    const ret = await api.post('user/logout');
    await $.ret(ret);
    if (ret.code === 1) {
      history.push(app.url('admin/login', {next: window.location.pathname}));
    }
  }

  render() {
    const {user} = this.props;

    return (
      <Box
        as={Layout.Header}
        bg="white"
        sx={{
          boxShadow: 'sm',
          zIndex: 1,
        }}
      >
        <Dropdown
          className={css({float: 'right'})}
          overlay={
            <Menu>
              <Menu.Item>
                <Link to={app.url('admin/password')}>
                  <i className="fa fa-lock"/>{' '}修改密码
                </Link>
              </Menu.Item>
              <Menu.Item>
                <a onClick={this.handleLogout}>
                  <i className="fa fa-power-off"/>{' '}退出
                </a>
              </Menu.Item>
            </Menu>
          }
        >
          <a className="ant-dropdown-link" onClick={e => e.preventDefault()}>
            <Actions>
              <Avatar src={user.avatar}>
                {user.username ? user.username.substr(0, 1).toUpperCase() : ''}
              </Avatar>
              {user.username}
              <DownOutlined/>
            </Actions>
          </a>
        </Dropdown>
      </Box>
    );
  }
}
