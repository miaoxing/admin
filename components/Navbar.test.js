import Navbar from './Navbar';
import {fireEvent, render, waitFor} from '@testing-library/react';
import {history} from '@mxjs/app';
import $, {Ret} from 'miaoxing';
import {MemoryRouter} from 'react-router';
import {bootstrap} from '@mxjs/test';

bootstrap();

describe('Navbar', () => {
  test('logout', async () => {
    $.http = jest.fn().mockResolvedValueOnce({
      ret: Ret.suc(),
    });

    history.push = jest.fn();

    const result = render(<MemoryRouter>
      <Navbar user={{
        username: '测试用户名',
      }}/>
    </MemoryRouter>);

    fireEvent.mouseOver(result.getByText('测试用户名'));

    const logout = await result.findByText('退出登录');
    fireEvent.click(logout);

    await waitFor(() => {
      expect(history.push).toBeCalled();
    });

    expect($.http).toMatchSnapshot();
    expect(history.push).toMatchSnapshot();
  });
});
