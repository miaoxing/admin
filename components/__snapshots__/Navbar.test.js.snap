// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar logout 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "post",
        "url": "/api/logout",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Navbar logout 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "/admin/login?next=%2F",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
